data_safe_haven/commands/config.py

```diff
@@ -68,18 +86,74 @@ def upload(
     config.upload(context)
 
 
+# Commands related to an SRE
 @config_command_group.command()
-def show(
+def show_sre(
+    name: Annotated[str, typer.Argument(help="Name of SRE to show")],
     file: Annotated[
         Optional[Path],  # noqa: UP007
         typer.Option(help="File path to write configuration template to."),
-    ] = None
+    ] = None,
 ) -> None:
-    """Print the configuration for the selected Data Safe Haven context"""
+    """Print the SRE configuration for the selected SRE and Data Safe Haven context"""
     context = ContextSettings.from_file().assert_context()
-    config = Config.from_remote(context)
+    sre_config = SREConfig.from_remote_by_name(context, name)
+    config_yaml = sre_config.to_yaml()
     if file:
         with open(file, "w") as outfile:
-            outfile.write(config.to_yaml())
+            outfile.write(config_yaml)
     else:
-        rprint(config.to_yaml())
+        rprint(config_yaml)
+
+
+@config_command_group.command()
+def template_sre(
+    file: Annotated[
+        Optional[Path],  # noqa: UP007
+        typer.Option(help="File path to write configuration template to."),
+    ] = None
+) -> None:
+    """Write a template Data Safe Haven SRE configuration."""
+    sre_config = SREConfig.template()
+    # The template uses explanatory strings in place of the expected types.
+    # Serialisation warnings are therefore suppressed to avoid misleading the users into
+    # thinking there is a problem and contaminating the output.
+    config_yaml = sre_config.to_yaml(warnings=False)
+    if file:
+        with open(file, "w") as outfile:
+            outfile.write(config_yaml)
+    else:
+        rprint(config_yaml)
+
+
+@config_command_group.command()
+def upload_sre(
+    file: Annotated[Path, typer.Argument(help="Path to configuration file")],
+) -> None:
+    """Upload an SRE configuration to the Data Safe Haven context"""
+    context = ContextSettings.from_file().assert_context()
+    logger = get_logger()
+
+    # Create configuration object from file
+    with open(file) as config_file:
+        config_yaml = config_file.read()
+    config = SREConfig.from_yaml(config_yaml)
+
+    # Present diff to user
+    if SREConfig.remote_exists(context, filename=config.filename):
+        if diff := config.remote_yaml_diff(context, filename=config.filename):
+            for line in "".join(diff).splitlines():
+                logger.info(line)
+            if not prompts.confirm(
+                (
+                    "Configuration has changed, "
+                    "do you want to overwrite the remote configuration?"
+                ),
+                default_to_yes=False,
+            ):
+                raise typer.Exit()
+        else:
+            rprint("No changes, won't upload configuration.")
+            raise typer.Exit()
+
+    config.upload(context, filename=config.filename)
```

Comment:
Maybe this part could be shared, assuming `SREConfig` and `SHMConfig` have the same methods.

Reply:
I see now how they are actually a bit different due to the filename argument.

Reply:
Yeah, it's a bit annoying. You can't use a class-level variable to hold the filename (as it's instance-dependent) but we still need to be able to load with a class-level constructor. This leaves you with a few `object.class_method(object.instance_attribute)` calls.